python/cudf/cudf/core/indexed_frame.py

```diff
@@ -414,3 +415,60 @@ def sort_index(
         if ignore_index is True:
             out = out.reset_index(drop=True)
         return self._mimic_inplace(out, inplace=inplace)
+
+    def _gather(self, gather_map, keep_index=True, nullify=False):
+        """Gather rows of frame specified by indices in `gather_map`.
+
+        Set rows to null for all out of bound indices if nullify is `True`.
+        """
+        gather_map = cudf.core.column.as_column(gather_map)
+        if not is_integer_dtype(gather_map.dtype):
+            gather_map = gather_map.astype("int32")
+
+        if not nullify and not _gather_map_is_valid(gather_map, len(self)):
+            raise IndexError("Gather map index is out of bounds.")
+
+        result = self.__class__._from_maybe_indexed_columns(
+            libcudf.copying.gather(
+                list(self._index._columns + self._columns)
+                if keep_index
+                else list(self._columns),
+                gather_map,
+                nullify=nullify,
+            ),
+            self._column_names,
+            self._index.names if keep_index else None,
+        )
+
+        result._copy_type_metadata(self, include_index=keep_index)
+        if keep_index:
+            result.index.names = self._index.names
+        return result
+
+    def _positions_from_column_names(self, column_names, include_index=False):
+        """Map each column name into their positions in the frame.
+
+        Return positions of the provided column names, offset by the number of
+        index columns if index exists and `include_index` is True. The order
+        of indices returned corresponds to the column order in this Frame.
+        """
+        n_indices_cols = len(self._index._data) if include_index else 0
+        return [
+            i + n_indices_cols
```

Comment:
Minor nit: `num_index_columns` might be better. We don't have plural "indices," just potentially a multi-index. 😉
```suggestion
        num_index_columns = len(self._index._data) if include_index else 0
        return [
            i + num_index_columns
```